### PR TITLE
fix: git branch protection bug fixes

### PR DESCRIPTION
**Pull request by brayn003**
## Description
Fixes a couple of issues with branch protection
- #28727 Toast message issues
- #28726 Pull count on pull icon
- #28748 Display remote branches in list
- #28697 Pulling protected branch shows discard toast
- Added branch name to default branch toast
- #28698 Select not scrolling bug

#### PR fixes following issue(s)
Mentioned above

#### Type of change
- Bug fix (non-breaking change which fixes an issue)
>
>
>
- [x] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx [hunk 1]
```diff
@@ -56,13 +56,33 @@ function GitProtectedBranches() {
   const dispatch = useDispatch();
 
   const unfilteredBranches = useSelector(getGitBranches);
-  const branches = unfilteredBranches.filter(
-    (b) => !b.branchName.includes("origin/"),
-  );
+  const defaultBranch = useSelector(getDefaultGitBranchName);
+
+  const brancheNames = useMemo(() => {
```
Comment (dipyamanbiswas07) [inline app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx:61]: Spelling

File: app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx [hunk 2]
```diff
@@ -56,13 +56,33 @@ function GitProtectedBranches() {
...
+    const remoteBrancheNames = [];
```
Comment (dipyamanbiswas07) [inline app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx:62]: Spelling

File: app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx [hunk 3]
```diff
@@ -56,13 +56,33 @@ function GitProtectedBranches() {
...
+    const localBranchNames = [];
+    for (const unfilteredBranch of unfilteredBranches) {
+      if (unfilteredBranch.branchName === defaultBranch) {
+        continue;
+      }
+      if (unfilteredBranch.branchName.includes("origin/")) {
```
Comment (dipyamanbiswas07) [inline app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx:68]: Should we store /origin in a constant? We should not use magic strings

File: app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx (hunk 1 above)
Comment (dipyamanbiswas07) [reply app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx:61]: brancheNames

File: app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx (hunk 2 above)
Comment (dipyamanbiswas07) [reply app/client/src/pages/Editor/gitSync/Tabs/GitSettings/GitProtectedBranches.tsx:62]: remoteBrancheNames

File: app/client/src/sagas/GitSyncSagas.ts [hunk 4]
```diff
@@ -1040,10 +1044,15 @@ function* discardChanges() {
     );
     if (isValidResponse) {
       yield put(discardChangesSuccess(response.data));
-      // const applicationId: string = response.data.id;
+      const successToastMessage =
+        payload?.successToastMessage ?? createMessage(DISCARD_SUCCESS);
+      toast.show(successToastMessage, {
```
Comment (sneha122) [inline app/client/src/sagas/GitSyncSagas.ts:1049]: @brayn003 If successToastMessage is undefined, will it result in blank toast message on screen? If yes we should maybe enclose this in a if block and show toast only when message is defined

**Review by dipyamanbiswas07 — DISMISSED** [review]
Lgtm

File: app/client/src/sagas/GitSyncSagas.ts (hunk 4 above)
Comment (brayn003) [reply app/client/src/sagas/GitSyncSagas.ts:1049]: successToastMessage will never be undefined, it will always default to `createMessage(DISCARD_SUCCESS)`

**Review by dipyamanbiswas07 — APPROVED** [review]

**Pull request merged**